### PR TITLE
Add char-based language model training process for aishell.

### PR DESCRIPTION
**Pull request by li563042811**
Add prepare_char_lm_training_data.py in local/ for char-based lm train data preparation process.
Add steps 9,10,11 in prepare.sh for lm data preparation and RNN lm training.
https://github.com/k2-fsa/icefall/issues/920

### CODE REVIEW

**Review by marcoyang1998 — APPROVED** [review]

**Comment by marcoyang1998** [comment]
Thanks for your contribution!

Do you have any WERs to report with LM fusion? It would be great to have some numbers here.

File: egs/aishell/ASR/prepare.sh [hunk 1]
```diff
@@ -219,3 +219,91 @@ if [ $stage -le 8 ] && [ $stop_stage -ge 8 ]; then
...
+  log "Stage 10: Sort LM training data"
+  # Sort LM training data by sentence length in descending order
+  # for ease of training.
+  #
+  # Sentence length equals to the number of tokens
+  # in a sentence.
+
+  out_dir=data/lm_training_char
+  mkdir -p $out_dir
+  ./local/sort_lm_training_data.py \
```
Comment (marcoyang1998) [inline egs/aishell/ASR/prepare.sh:277]: Could you add a soft link for this file? I don't think it's present under `local`

**Review by marcoyang1998 — CHANGES_REQUESTED** [review]

File: egs/aishell/ASR/prepare.sh (hunk 1 above)
Comment (li563042811) [reply egs/aishell/ASR/prepare.sh:277]: Ok, I will add it as soon as possible.

**Review by marcoyang1998 — APPROVED** [review]

**Comment by marcoyang1998** [comment]
LGTM!

Merging it now.

**Comment by minmin0** [comment]
good job! Can we use SF in sherpa-onnx now? I found kaldi and  wenet already support LM in C++  runtime, but in sherpa-onnx or sherpa-ncnn  i don't know how to use extern LM.

**Comment by marcoyang1998** [comment]
I would recommend using RNNLM rescoring in #1002 in Sherpa as it is much faster than shallow fusion while yielding similar WERs. This is on our to-do list.

**Comment by minmin0** [comment]
> RNNLM

Thanks, but the decoding code in icefall is not suitable for use in the production environment.I think sherp-onnx maybe one option, but unfortunately, it is not supported LM.

**Comment by csukuangfj** [comment]
> > RNNLM
> 
> Thanks, but the decoding code in icefall is not suitable for use in the production environment.I think sherp-onnx maybe one option, but unfortunately, it is not supported LM.

@minmin0 Can you help to test it in sherpa-onnx if it is supported in sherpa-onnx?

**Comment by minmin0** [comment]
> onnx

yes

**Comment by minmin0** [comment]
> > > RNNLM
> > 
> > 
> > Thanks, but the decoding code in icefall is not suitable for use in the production environment.I think sherp-onnx maybe one option, but unfortunately, it is not supported LM.
> 
> @minmin0 Can you help to test it in sherpa-onnx if it is supported in sherpa-onnx?

yes， I am willing to do  that

**Pull request merged**